Add optional_host_permissions docs

rebloor opened the pull request:
### Description

Adds documentation for `optional_host_permissions` in support of [Bug 1766026](https://bugzilla.mozilla.org/show_bug.cgi?id=1766026) Add optional_host_permissions in mv3 including:

- new `optional_host_permissions` key page
- updates to `optional_permissions` key page re specification of hosts
- update to `permissions` key page host permissions section to explain the various options for requesting host permissions (given that this is the de facto reference point for host permissions information)
- release note

### Related issues and pull requests

Related BCD changes in https://github.com/mdn/browser-compat-data/pull/23177

rebloor committed:
Add optional_host_permissions docs

rebloor committed:
Linter suggestion

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

rebloor committed:
Merge branch 'main' into Add-optional_host_permissions-docs

zombie commented on files/en-us/mozilla/add-ons/webextensions/api/permissions/contains/index.md:
```diff
@@ -9,7 +9,7 @@ browser-compat: webextensions.api.permissions.contains
 Check whether the extension has the permissions listed in the given {{WebExtAPIRef("permissions.Permissions")}} object.
 
-The `Permissions` argument may contain either an origins property, which is an array of [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions), or a `permissions` property, which is an array of [API permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#api_permissions), or both.
+The `Permissions` argument may contain either an origins property, which is an array of [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions), [optional_host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#optional_host_permissions), or a `permissions` property, which is an array of [API permissions](/en-US/docs/M ...
```
I don't see where you've added the `#optional_host_permissions` anchor to the `manifesst.json/permissions` page, did I miss something?

In any case, I don't think we need this and the next three changes to the API methods docs, the existing text already links to the general explanation about host permissions [1], not to a specific manifest key like `optional_permissions`, so linking to `optional_host_permissions` key doesn't seem necessary here IMO.

rebloor replied:
@zombie Good point - I think I got carried away there! On further review, I also see that I needed a change under that "Host permissions" section

rebloor replied:
```suggestion
The `Permissions` argument may contain either an origins property, which is an array of [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions) or a `permissions` property, which is an array of [API permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#api_permissions), or both.
```

rebloor commented on files/en-us/mozilla/add-ons/webextensions/manifest.json/optional_permissions/index.md:
```diff
@@ -39,18 +39,15 @@ The [`permissions`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/perm
 For advice on designing your request for runtime permissions, to maximize the likelihood that users grant them, see [Request permissions at runtime](https://extensionworkshop.com/documentation/develop/request-the-right-permissions/#request_permissions_at_runtime).
 
-Starting with Firefox 84, users will be able to manage optional permissions from the Firefox Add-ons Manager. Extensions that use optional permissions should listen for [browser.permissions.onAdded](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/permissions/onAdded) and [browser.permissions.onRemoved](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/permissions/onRemoved) API events to know when a us ...
+Starting with Firefox 84, users can to manage optional permissions from the Firefox Add-ons Manager. Extensions that use optional permissions can listen for [browser.permissions.onAdded](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/permissions/onAdded) and [browser.permissions.onRemoved](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/permissions/onRemoved) API events to know when a user grants or ...
```
```suggestion
Starting with Firefox 84, users can manage optional permissions from the Firefox Add-ons Manager. Extensions that use optional permissions can listen for [browser.permissions.onAdded](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/permissions/onAdded) and [browser.permissions.onRemoved](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/permissions/onRemoved) API events to know when a user grants or revokes these permissions.
```

rebloor commented on files/en-us/mozilla/add-ons/webextensions/api/permissions/index.md:
```diff
@@ -20,7 +20,7 @@ With the permissions API, an extension can ask for additional permissions at run
 - The extension can run with a smaller set of permissions, except when it actually needs them.
 - The extension can handle permission denial in a graceful manner, instead of presenting the user with a global "all or nothing" choice at install time. You can still get a lot out of that map extension, without giving it access to your location, for example.
-- The extension may need [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions), but not know at install time which host permissions it needs. For example, the list of hosts may be a user setting. In this scenario, asking for a more specific range of hosts at runtime, can be an alternative to asking for "\<all_urls>" at install time.
+- The extension may need [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions) or [optional_host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#optional_host_permissions), but not know at install time which host permissions it needs. For example, the list of hosts may be a user setting. In this scenario, asking f ...
```
```suggestion
- The extension may need [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions), but not know at install time which host permissions it needs. For example, the list of hosts may be a user setting. In this scenario, asking for a more specific range of hosts at runtime, can be an alternative to asking for "\<all_urls>" at install time.
```

rebloor commented on files/en-us/mozilla/add-ons/webextensions/api/permissions/permissions/index.md:
```diff
@@ -14,7 +14,7 @@ A `Permissions` object represents a collection of permissions.
 
 - `origins` {{optional_inline}}
-  - : An array of [match patterns](/en-US/docs/Mozilla/Add-ons/WebExtensions/Match_patterns), representing [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions).
+  - : An array of [match patterns](/en-US/docs/Mozilla/Add-ons/WebExtensions/Match_patterns), representing [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions) or [optional_host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#optional_host_permissions).
```
```suggestion
  - : An array of [match patterns](/en-US/docs/Mozilla/Add-ons/WebExtensions/Match_patterns), representing [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions).
```

rebloor commented on files/en-us/mozilla/add-ons/webextensions/api/permissions/remove/index.md:
```diff
@@ -9,7 +9,7 @@ browser-compat: webextensions.api.permissions.remove
 Ask to give up the permissions listed in the given {{WebExtAPIRef("permissions.Permissions")}} object.
 
-The `Permissions` argument may contain either an `origins` property, which is an array of [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions), or a `permissions` property, which is an array of [API permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#api_permissions), or both. Permissions must come from the set of pe ...
+The `Permissions` argument may contain either an `origins` property, which is an array of [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions), [optional_host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#optional_host_permissions),or a `permissions` property, which is an array of [API permissions](/en-US/docs/ ...
```
```suggestion
The `Permissions` argument may contain either an `origins` property, which is an array of [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions), or a `permissions` property, which is an array of [API permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#api_permissions), or both. Permissions must come from the set of permissions defined in the [`optional_permissions`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/optional_permissions) manifest.json key.
```

rebloor commented on files/en-us/mozilla/add-ons/webextensions/api/permissions/request/index.md:
```diff
@@ -9,7 +9,7 @@ browser-compat: webextensions.api.permissions.request
 Ask for the set of permissions listed in the given {{WebExtAPIRef("permissions.Permissions")}} object.
 
-The `Permissions` argument may contain either an `origins` property, which is an array of [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions), or a `permissions` property, which is an array of [API permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#api_permissions), or both. Permissions must come from the set of pe ...
+The `Permissions` argument may contain either an `origins` property, which is an array of [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions), [optional_host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#optional_host_permissions), or a `permissions` property, which is an array of [API permissions](/en-US/docs ...
```
```suggestion
The `Permissions` argument may contain either an `origins` property, which is an array of [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions), or a `permissions` property, which is an array of [API permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#api_permissions), or both. Permissions must come from the set of permissions defined in the [`optional_permissions`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/optional_permissions) manifest.json key. The `origins` property may include permissions that match a subset of the hosts matched by an optional permission: for example, if optional_permissions include "\*://mozilla.org/", then `permissions.origins` may include "https\://developer.mozilla.org/".
```

rebloor committed:
Apply suggestions from review

rebloor commented on files/en-us/mozilla/add-ons/webextensions/api/permissions/contains/index.md:
```diff
@@ -9,7 +9,7 @@ browser-compat: webextensions.api.permissions.contains
 Check whether the extension has the permissions listed in the given {{WebExtAPIRef("permissions.Permissions")}} object.
 
-The `Permissions` argument may contain either an origins property, which is an array of [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions), or a `permissions` property, which is an array of [API permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#api_permissions), or both.
+The `Permissions` argument may contain either an origins property, which is an array of [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions) or a `permissions` property, which is an array of [API permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#api_permissions), or both.
```
```suggestion
The `Permissions` argument may contain either an origins property, which is an array of [host permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions), or a `permissions` property, which is an array of [API permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#api_permissions), or both.
```

rebloor committed:
Corection to api/permissions/contains

rebloor committed:
additional xref
Addition of link to optional_host_permissions in the host permission section of the permissions manifest key page

rebloor committed:
Revert "additional xref"

This reverts commit ea61df050201b54c63eccfed60c52f88c95ab55c.

rebloor committed:
Expanded info in permissions/host permissions

rebloor committed:
Apply linter suggestion

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

rebloor committed:
Merge branch 'main' into Add-optional_host_permissions-docs

zombie reviewed (approved):
thanks, lgtm.

dotproto commented on files/en-us/mozilla/add-ons/webextensions/manifest.json/optional_host_permissions/index.md:
```diff
@@ -0,0 +1,54 @@
+
+See [`host_permissions`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/host_permissions) for more information on defining hosts.
+
+Use the {{webextapiref("permissions")}} API to ask for a permission. Asking for a permission may present the user with a dialog requesting they grant the host permission to your extension.
```
Since the method that the extension would call is named `request()`, I think we should use that verb for the operation performed by the developer rather than "ask."

```suggestion
Use the {{webextapiref("permissions")}} API to request an optional host permission at runtime. Requesting a host permission may present the user with a dialog asking whether they want to grant the permission to your extension.
```

rebloor committed:
Merge branch 'main' into Add-optional_host_permissions-docs

dotproto reviewed (approved):
Only minor changes requested. Pre-approving to avoid blocking.

rebloor committed:
Review feedback

Co-authored-by: Simeon Vincent <svincent@gmail.com>
